File: TN-POS-SYS/TN-SYS-WEB/src/core/api/api_client.ts
// 🇻🇳 Client API cơ sở
// 🇺🇸 Base API client
import type { M_Api_Response } from './m_api_response';

export class Api_Client {
  private baseUrl: string;

  constructor(baseUrl?: string) {
    // 🇻🇳 Sử dụng proxy trong development hoặc baseUrl được cung cấp
    // 🇺🇸 Use proxy in development or provided baseUrl
    if (baseUrl) {
      this.baseUrl = baseUrl;
    } else if (typeof window !== 'undefined') {
      // Trong development, sử dụng proxy /api
      // In production, có thể cấu hình qua environment variable
      const apiBase = import.meta.env.VITE_API_BASE_URL || '/api';
      this.baseUrl = apiBase.startsWith('http') ? apiBase : `${window.location.origin}${apiBase}`;
    } else {
      this.baseUrl = '';
    }
  }

  // 🇻🇳 Gửi request GET
  // 🇺🇸 Send GET request
  async get<T>(path: string, options?: RequestInit): Promise<M_Api_Response<T>> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
        ...options?.headers,
      },
      ...options,
    });

    return this.handleResponse<T>(response);
  }

  // 🇻🇳 Gửi request POST
  // 🇺🇸 Send POST request
  async post<T>(path: string, body?: unknown, options?: RequestInit): Promise<M_Api_Response<T>> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
        ...options?.headers,
      },
      body: body ? JSON.stringify(body) : undefined,
      ...options,
    });

    return this.handleResponse<T>(response);
  }

  // 🇻🇳 Gửi request DELETE
  // 🇺🇸 Send DELETE request
  async delete<T>(path: string, options?: RequestInit): Promise<M_Api_Response<T>> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders(),
        ...options?.headers,
      },
      ...options,
    });

    return this.handleResponse<T>(response);
  }

  // 🇻🇳 Lấy headers xác thực
  // 🇺🇸 Get authentication headers
  private getAuthHeaders(): Record<string, string> {
    const token = this.getToken();
    return token ? { Authorization: `Bearer ${token}` } : {};
  }

  // 🇻🇳 Lấy token từ storage
  // 🇺🇸 Get token from storage
  private getToken(): string | null {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem('auth_token');
  }

  // 🇻🇳 Xử lý phản hồi
  // 🇺🇸 Handle response
  private async handleResponse<T>(response: Response): Promise<M_Api_Response<T>> {
    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: response.statusText }));
      throw new Error(error.message || `HTTP ${response.status}`);
    }

    return response.json();
  }
}

File: TN-POS-SYS/TN-SYS-FLU/src/core/api/api_client.dart
// 🇻🇳 Client API cơ sở
// 🇺🇸 Base API client
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'm_api_response.dart';

class Api_Client {
  final String baseUrl;

  Api_Client({String? baseUrl})
      : baseUrl = baseUrl ?? 'http://localhost:3000';

  // 🇻🇳 Lấy token từ storage
  // 🇺🇸 Get token from storage
  String? _getToken() {
    // TODO: Implement token retrieval from secure storage
    return null;
  }

  // 🇻🇳 Lấy headers xác thực
  // 🇺🇸 Get authentication headers
  Map<String, String> _getAuthHeaders() {
    final token = _getToken();
    return token != null ? {'Authorization': 'Bearer $token'} : {};
  }

  // 🇻🇳 Gửi request GET
  // 🇺🇸 Send GET request
  Future<M_Api_Response<T>> get<T>(
    String path, {
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final response = await http.get(
      Uri.parse('$baseUrl$path'),
      headers: {
        'Content-Type': 'application/json',
        ..._getAuthHeaders(),
      },
    );

    return _handleResponse<T>(response, fromJson);
  }

  // 🇻🇳 Gửi request POST
  // 🇺🇸 Send POST request
  Future<M_Api_Response<T>> post<T>(
    String path, {
    Object? body,
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl$path'),
      headers: {
        'Content-Type': 'application/json',
        ..._getAuthHeaders(),
      },
      body: body != null ? jsonEncode(body) : null,
    );

    return _handleResponse<T>(response, fromJson);
  }

  // 🇻🇳 Gửi request DELETE
  // 🇺🇸 Send DELETE request
  Future<M_Api_Response<T>> delete<T>(
    String path, {
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final response = await http.delete(
      Uri.parse('$baseUrl$path'),
      headers: {
        'Content-Type': 'application/json',
        ..._getAuthHeaders(),
      },
    );

    return _handleResponse<T>(response, fromJson);
  }

  // 🇻🇳 Xử lý phản hồi
  // 🇺🇸 Handle response
  Future<M_Api_Response<T>> _handleResponse<T>(
    http.Response response,
    T Function(Map<String, dynamic>)? fromJson,
  ) async {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final apiResponse = M_Api_Response.fromJson(
        json,
        fromJson != null ? (obj) => fromJson(obj as Map<String, dynamic>) : null,
      );
      return apiResponse;
    } else {
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    }
  }
}

